File: app/src/main/java/org/antrack/app/ui/fragments/ModulesAdapter.java
package org.antrack.app.ui.fragments;

import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.antrack.app.ui.Module;
import org.antrack.app.ui.RecyclerViewAnim;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;

import app.R;

public class ModulesAdapter extends RecyclerViewAnim.Adapter<ModulesAdapter.ModuleViewHolder> {
    public static class ModuleViewHolder extends RecyclerView.ViewHolder {
        CardView cv;
        TextView moduleName;
        TextView moduleDesc;
        TextView moduleVersion;
        TextView moduleAuthor;

        ModuleViewHolder(View itemView) {
            super(itemView);
            cv = (CardView)itemView.findViewById(R.id.cardview_modules);
            moduleName = (TextView)itemView.findViewById(R.id.cardview_modules_name);
            moduleDesc = (TextView)itemView.findViewById(R.id.cardview_modules_desc);
            moduleVersion = (TextView)itemView.findViewById(R.id.cardview_modules_version);
            moduleAuthor = (TextView)itemView.findViewById(R.id.cardview_modules_author);
        }
    }

    LinkedHashMap<String, Module> modules;

    ModulesAdapter(LinkedHashMap<String, Module> modules){
        this.modules = modules;
    }

    public void update(LinkedHashMap<String, Module> modules) {
        this.modules = modules;
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public ModuleViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.cardview_modules, viewGroup, false);
        return new ModuleViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ModuleViewHolder viewHolder, int i) {
        Module module = (Module)getElementByIndex(modules, i);
        viewHolder.moduleName.setText(module.name);
        viewHolder.moduleDesc.setText(module.desc);
        viewHolder.moduleVersion.setText("Version: " + module.version);
        viewHolder.moduleAuthor.setText("Author: " + module.author);
        // FIXME disabled: причина, кнопка launch
    }

    private Object getElementByIndex(LinkedHashMap map, int index){
        return map.get(map.keySet().toArray()[index]);
    }

    @Override
    public int getItemCount() {
        return modules.size();
    }

}

File: app/src/main/java/org/antrack/app/ui/fragments/ModulesAdapter.kt
package org.antrack.app.ui.fragments

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import app.R
import org.antrack.app.ui.Module
import java.util.*

class ModulesAdapter internal constructor(internal var modules: LinkedHashMap<String, Module>) : RecyclerView.Adapter<ModulesAdapter.ModuleViewHolder>() {
    class ModuleViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal var moduleName: TextView = itemView.findViewById(R.id.cardview_modules_name) as TextView
        internal var moduleDesc: TextView = itemView.findViewById(R.id.cardview_modules_desc) as TextView
        internal var moduleVersion: TextView = itemView.findViewById(R.id.cardview_modules_version) as TextView
        internal var moduleAuthor: TextView = itemView.findViewById(R.id.cardview_modules_author) as TextView
    }

    fun update(modules: LinkedHashMap<String, Module>) {
        this.modules = modules
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        super.onAttachedToRecyclerView(recyclerView)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ModuleViewHolder {
        val v = LayoutInflater.from(viewGroup.context).inflate(R.layout.cardview_modules, viewGroup, false)
        return ModuleViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ModuleViewHolder, i: Int) {
        val module = getElementByIndex(modules, i) as Module
        viewHolder.moduleName.text = module.name
        viewHolder.moduleDesc.text = module.desc
        viewHolder.moduleVersion.text = "Version: " + module.version!!
        viewHolder.moduleAuthor.text = "Author: " + module.author!!
    }

    private fun getElementByIndex(map: LinkedHashMap<*, *>, index: Int): Any? {
        return map[map.keys.toTypedArray()[index]]
    }

    override fun getItemCount(): Int {
        return modules.size
    }

}
